### PR TITLE
feat: add `Foldable.joinWith`

### LINKED ISSUE #3047: Add `Foldable.joinWith` and `X.joinWith`
Add 

```scala
pub def joinWith(f: a -> String, sep: String, f: f[a]): String = ...
```

to `Foldable`. 

Add

```scala
pub def joinWith(f: a -> String, sep: String, l: List[a]): String = ...
```

and similar to each collection namespace. (These can just call Foldable, except for mutable data structures which must be impure).

In each case , the implementation should use a `StringBuilder` for efficiency. (Note this requires a cast)

### PR DESCRIPTION
**Pull request by jaschdoc**
Closes #3047

- [x]  `List`
- [x]  `Nel`
- [x]  `Mutlist`
- [x]  `LazyList`
- [x]  `DemandList`
- [x]  `Array`
- [x]  `Iterator`
- [x]  `RedBlackTree`
- [x]  `Map`
- [x]  `DelayMap`
- [x]  `MutMap`
- [x]  `Set`
- [x]  `MutSet`
- [x]  `Chain`
- [x]  `MutDeque`

### CODE REVIEW

**Comment by jaschdoc** [comment]
@magnus-madsen RFR! Should we consider removing `joinWithValues` + `joinWithKeys` and simply have `joinWith` in `RedBlackTree` and friends?

**Comment by magnus-madsen** [comment]
> @magnus-madsen RFR! Should we consider removing `joinWithValues` + `joinWithKeys` and simply have `joinWith` in `RedBlackTree` and friends?

What do you mean more precisely? What am I choosing between and what do you recommend?

**Comment by jaschdoc** [comment]
I've added 3 `joinWith` functions for `RedBlackTree` and friends: `joinWith`, `joinWithValues`, and `joinWithKeys`. What do you think of this? I suggest removing the latter two and only have `joinWith`

**Comment by magnus-madsen** [comment]
> I've added 3 `joinWith` functions for `RedBlackTree` and friends: `joinWith`, `joinWithValues`, and `joinWithKeys`. What do you think of this? I suggest removing the latter two and only have `joinWith`

OK. I agree with this. One joinWith is enough!

**Comment by jaschdoc** [comment]
@magnus-madsen ready for review

File: main/src/library/Array.flix [hunk 1]
```diff
@@ -1614,9 +1614,17 @@ namespace Array {
...
+    ///
+    /// Returns the concatenation of the string representation
+    /// of each element in `a` according to `f` with `sep` inserted between each element.
+    ///
+    pub def joinWith(f: a -> String & ef, sep: String, a: Array[a]): String & Impure =
+        let lastSep = String.length(sep);
         let sb = StringBuilder.new();
-        foldLeft((_, x) -> StringBuilder.append!(sb, "${x}" + sep), (), a);
-        StringBuilder.toString(sb) |> String.dropRight(1)
+        foldLeft((_, x) -> StringBuilder.append!(sb, "${f(x)}" + sep), (), a);
```
Comment (magnus-madsen) [inline main/src/library/Array.flix:1626]: Its better (for performance) to call append! twice than to use +.

Comment (magnus-madsen) [reply main/src/library/Array.flix:1626]: (This may apply in other places too?)

File: main/src/library/Foldable.flix [hunk 2]
```diff
@@ -294,9 +294,21 @@ pub lawless class Foldable[t : Type -> Type] {
...
+    } as & Pure
+
+    ///
+    /// Returns the concatenation of the string representation
+    /// of each element in `t` according to `f` with `sep` inserted between each element.
+    ///
+    pub def joinWith(f: a -> String & ef, sep: String, t: t[a]): String = {
+        let lastSep = String.length(sep);
+        let sb = StringBuilder.new();
+        t |> Foldable.foldLeft((_, x) -> StringBuilder.append!(sb, "${f(x)}" + sep), ());
```
Comment (magnus-madsen) [inline main/src/library/Foldable.flix:310]: Same comment about avoiding +

File: main/src/library/Array.flix (hunk 1 above)
Comment (jaschdoc) [reply main/src/library/Array.flix:1626]: Roger

**Comment by jaschdoc** [comment]
Should be fixed now

**Comment by magnus-madsen** [comment]
Thanks!

**Pull request merged**